v2/apierror/apierror.go

```diff
@@ -60,6 +61,23 @@ type ErrDetails struct {
 	Unknown []interface{}
 }
 
+// ExtractMessage provides a mechanism for extracting messages from the Unknown
+// error details.  An entry matching the type of exampleMesg is returned
+// if present.
+func (e ErrDetails) ExtractMessage(exampleMesg interface{}) interface{} {
+	if exampleMesg == nil {
+		return nil
+	}
+	typ := reflect.TypeOf(exampleMesg)
+	for _, elem := range e.Unknown {
+		elemType := reflect.TypeOf(elem)
```

Comment:
Would it be possible/better to use `protoreflect` here instead? And maybe for the type(s) in the method signature to be narrowed to a proto type as well?

Reply:
Avoiding the use of protoreflect here as it's not clear this will always be proto messages, such as in the case of rest transports.

Reply:
> Avoiding the use of protoreflect here as it's not clear this will always be proto messages, such as in the case of rest transports

The details in google API errors will always be a google.protobuf.Any, which means it will always be a protobuf message stuffed in there. i think it is safe to assume this will always be a proto message. [Docs](https://cloud.google.com/apis/design/errors#http_mapping).

Did you have another case in mind?

Reply:
Currently this PR adds this method: 
```func (e ErrDetails) ExtractMessage(v interface{}) error```

If we switch this to something protobuf-based, are you envisioning that surfacing in the public contract or just the implementation?

```func (e ErrDetails) ExtractMessage(md protoreflect.MessageDescriptor) (proto.Message, error)```
or 
```func (e ErrDetails) ExtractMessage(m protoreflect.Message) error```

These work, but still require the user to do extra work to get back to a statically defined custom error type published by a googleapi service.  In most (all?) cases that type should be a proto, so if they want to do something more exotic after the fact they can access proto reflection via their supplied type.

Reply:
Ok, so where I see it being a little beneficial is in the implementation. 

If the API takes a `proto.Message` (aliased from `protoreflect.Message`), and we type-cast the `Unknown` item to `proto.Message`, we can compare the names from [`proto.MessageName`](https://pkg.go.dev/google.golang.org/protobuf@v1.28.1/proto#MessageName), then do a [`proto.Merge`](https://pkg.go.dev/google.golang.org/protobuf@v1.28.1/proto#Merge). All protos in go-genproto should implement the `proto.Message` interface, so it _should_ just work ?

So I guess I misled you in saying "check out protoreflect". LMK what you think, and if that doesn't work, what is here will work.

Reply:
Swapped over to what I think you were envisioning (also renamed the method slightly to clarify the proto-based nature of the extraction).  PTAL.